spec/Classes.tex

```diff
@@ -784,12 +787,54 @@ \subsection{The Method Receiver and the {\em this} Argument}
 is \chpl{c1}.
 \end{chapelexample}
 
-The optional \sntx{this-intent} declares the intent by which the receiver
-argument will be passed.  When the value of \sntx{this-intent} is \chpl{param},
-it implies that the function can only be applied to param objects of the given
-type binding.  If the value is \chpl{ref}, the argument will be passed by
-reference---allowing modification of \chpl{this}.  Otherwise, the argument will
-be passed with the blank intent as specified in \rsec{The_Blank_Intent}.
+The optional \sntx{this-intent} declares how the receiver argument is
+constrained or will be passed for a given method call.
+
```

Comment:
On Jul 7, 2015, at 4:57 PM, Brad Chamberlain <notifications@github.com<mailto:notifications@github.com>> wrote:


In spec/Classes.tex<https://github.com/chapel-lang/chapel/pull/2144#discussion_r34104888>:

> @@ -784,12 +787,54 @@ \subsection{The Method Receiver and the {\em this} Argument}
>  is \chpl{c1}.
>  \end{chapelexample}
>
> -The optional \sntx{this-intent} declares the intent by which the receiver
> -argument will be passed.  When the value of \sntx{this-intent} is \chpl{param},
> -it implies that the function can only be applied to param objects of the given
> -type binding.  If the value is \chpl{ref}, the argument will be passed by
> -reference---allowing modification of \chpl{this}.  Otherwise, the argument will
> -be passed with the blank intent as specified in \rsec{The_Blank_Intent}.
> +The optional \sntx{this-intent} declares how the receiver argument is
> +constrained or will be passed for a given method call.
> +


If C++ permits static methods to be called on instances (which a quick test I wrote seems to indicate) then I think you're right that we shouldn't paint the analogy here (or else should change the implementation to match C++'s). I'd apparently completely forgotten (or never realized) that this is the case. Without looking, do you know whether purer OOP languages are similar in this respect? I think I learned the concept of methods on classes rather than instances from Smalltalk (?) and assumed it was the same in C++, but am now wondering if I misunderstood Smalltalk as well (I was a lazy student).



I am not too strong on Smalltalk but I think it’s tough/interesting to compare it with C++.

In particular in Smalltalk everything is an object.  Classes are just objects.  “By convention”, that doesn’t
seem quite right but .. , a primary responsibility of a Class object is to create other objects that have the same
type as each other.  In this model there’s no meaningful sense of static methods vs instance methods; there
are just methods.  Some happen to be attached to objects that inherit from class and others are attached to
objects that created by class objects.

Self and CLOS followed in the footsteps of Smalltalk in these regards.  In general it seems to me, that
Chapel is closer to Java/C++ than Smalltalk/CLOS.

I’ve know next to nothing about Java but I just did a quick Google and it appears to me that Java
follows C++; I found a reference that I believe says that one can invoke static methods (class methods)
via an instance or via the class name.


So I think we can agree that proc type methods in Chapel are currently different from Java/C++ static
methods.  I could go either way on whether Chapel should retain the current definition or follow C++.
One can argue that the Chapel definition is more pure but it has the obvious downside of being potentially
“surprising”.

Mike





—
Reply to this email directly or view it on GitHub<https://github.com/chapel-lang/chapel/pull/2144/files#r34104888>.